MonkeyLoader/Configuration/ConfigKeyRange.cs

```diff
@@ -0,0 +1,200 @@
+    /// </summary>
+    /// <remarks>
+    /// This class also implements <see cref="IConfigKeyValidator{T}"/> and is therefore two components in one.
+    /// </remarks>
+    /// <inheritdoc/>
+    public sealed class ConfigKeyRange<T> : IConfigKeyRange<T>, IConfigKeyValidator<T>
+    {
+        /// <inheritdoc/>
+        public IComparer<T?> Comparer { get; }
+
+        /// <inheritdoc/>
+        public T Max { get; }
+
+        /// <inheritdoc/>
+        public T Min { get; }
+
+        /// <summary>
+        /// Creates a new range component.
+        /// </summary>
+        /// <param name="min">The lower bound of the value range.</param>
+        /// <param name="max">The upper bound of the value range.</param>
+        /// <param name="comparer">The comparer to use to determine whether values fall into the range of this config item.</param>
+        /// <exception cref="ArgumentNullException">When <paramref name="min"/> or <paramref name="max"/> are null.</exception>
+        /// <exception cref="NotSupportedException">When <paramref name="comparer"/> is null while <typeparamref name="T"/> is not <see cref="IComparable{T}"/></exception>
+        public ConfigKeyRange(T? min = default, T? max = default, IComparer<T?>? comparer = null)
+        {
+            if (min is null)
+                throw new ArgumentNullException(nameof(min));
+
+            if (max is null)
+                throw new ArgumentNullException(nameof(max));
+
+            if (comparer is null && !typeof(IComparable<T>).IsAssignableFrom(typeof(T)))
+                throw new NotSupportedException($"The {nameof(comparer)} must not be null when {nameof(T)} is not IComparable<T>!");
+
+            Min = min;
+            Max = max;
+            Comparer = comparer ?? Comparer<T>.Default!;
+        }
+
+        /// <inheritdoc/>
+        public void Initialize(IDefiningConfigKey<T> config)
+        {
+            if (config.TryGetValue(out var value) && !IsValueInRange(value!))
+                throw new InvalidOperationException($"Value for key [{config.Id}] did not pass validation!");
+        }
+
+        /// <summary>
+        /// Determines whether the given value falls into this config item's range.
+        /// </summary>
+        /// <seealso cref="IsValueInRange(T)"/>
+        /// <param name="value">The value to test.</param>
+        /// <returns><c>true</c> if the value is not null and falls into the range; otherwise, <c>false</c>.</returns>
+        bool IConfigKeyValidator<T>.IsValid(T? value) => value is not null && IsValueInRange(value);
+
+        /// <inheritdoc/>
+        public bool IsValueInRange(T value)
+            => Comparer.Compare(Min, value) <= 0 && Comparer.Compare(Max, value) >= 0;
+    }
+
+    /// <summary>
+    /// Defines the typed definition for a ranged config item.
+    /// </summary>
+    /// <inheritdoc/>
+    public interface IConfigKeyRange<T> : IConfigKeyComponent<IDefiningConfigKey<T>>
```

Comment:
IMO it'd be nice to have this _before_ the actual implementation... Tho can't say I still even see the point of having this as an interface vs just using the actual impl type, I wouldn't expect anyone to DI this anyways... but this is c# so  guess that's just standard crazy levels and not weird

Reply:
Yea, the order is due to the automatic sorting 😅

I could feasibly see the Range one being implemented multiple times to split up comparer / IComparable using ones and stuff like that - however mainly the quantity one for resonite will use that too. Though it could be inheritance too.